joyent/mdb_v8#32: fix ::findjsobjects for V8 4.5.x

misterdjules opened the pull request:
This fixes ::findjsobjects for core dumps generated by Node.js 4.0.x
(and possibly later).

This change fixes how mdb_v8 retrieves the constructor of a JavaScript
object given its Map.

It also fixes the problem of accessing objects' properties that was
caused by a typo in "v8db_propindex_mask" that should have been
"v8db_prop_index_mask" (note the underscore in "prop_index") in my
previous changes.

It adds a "fallback" member for v8_offsets so that we can have a
fallback for typed arrays' length's offset.

Finally, this change allows ::findjsobjects to find Buffer instances and
inspect them. However, due to how Buffer instances are implemented in
node v4.x and later, they are currently seen by mdb_v8 as having a
constructor named "Uint8Array" instead of "Buffer", so ::findjsobjects
-c Buffer won't find any actual Buffer instances, instead one has to use
::findjsobjects -c Uint8Array.

misterdjules commented:
/cc @davepacheco

misterdjules commented:
Also, one thing I failed to mention in the commit message is that now Buffer instances don't have a property named `length`. Instead, the length is a generated property on typed arrays. The current changes print this generated property when we `::jsprint` a Buffer instance (really a typed array) to make the tests pass without having to change them too much.

However, I'm still undecided on whether we would want to display it in the final set of changes.

davepacheco commented on src/mdb_v8.c:
```diff
@@ -706,6 +721,11 @@ autoconfigure(v8_cfg_t *cfgp)
 		failed++;
 	}
 
+	if (V8_TYPE_JSTYPEDARRAY == -1) {
```
Is it definitely a problem for there to be no JSTypedArray type?  Even on older versions of Node?

misterdjules replied:
No, it's OK for older versions of node not to have `JSTypedArray`. Will fix too.

davepacheco commented on src/mdb_v8.c:
```diff
@@ -755,7 +775,8 @@ autoconfigure(v8_cfg_t *cfgp)
 		offset_added = offp->v8o_flags & V8_CONSTANT_ADDED;
-		v8_older = v8_version_older(v8_major, v8_minor, cnp->v8c_flags);
+		v8_older = v8_version_older(v8_major,
+		    v8_minor, offp->v8o_flags);
```
I don't really understand this change.  What were we passing before?

misterdjules replied:
We were passing the flags of the latest constant that had been processed (`cnp`), but what we really wanted to do is to pass the flags for the current v8_offset being processed.

davepacheco commented on src/mdb_v8.c:
```diff
@@ -1516,6 +1552,46 @@ read_heap_dict(uintptr_t addr,
 }
 
 /*
+ * Given a JavaScript object's Map "map", stores its constructor
```
Nit: this should probably say "the object's constructor", since it's not the Map's constructor.

misterdjules replied:
Absolutely, will update shortly.

davepacheco commented on src/mdb_v8.c:
```diff
@@ -1516,6 +1552,46 @@ read_heap_dict(uintptr_t addr,
+	 */
+	while (constructor_found == 0) {
+		if (read_heap_ptr(&constructor_candidate,
+			map, V8_OFF_MAP_CONSTRUCTOR) != 0)
```
nit: cstyle should complain about the continuation line here

misterdjules replied:
OK, `make prepush` didn't seem to complain. Will fix shortly.

davepacheco commented on src/mdb_v8.c:
```diff
@@ -2196,7 +2272,8 @@ jsobj_maybe_garbage(uintptr_t addr)
 	    type != V8_TYPE_JSDATE &&
-	    type != V8_TYPE_JSREGEXP)));
+	    type != V8_TYPE_JSREGEXP &&
+	    type != V8_TYPE_JSTYPEDARRAY)));
```
Do you know if the changing representation of Buffer has affected the "::nodebuffer" dcmd?

misterdjules replied:
See the changes in `dcmd_nodebuffer` below.

davepacheco commented on src/mdb_v8.c:
```diff
@@ -3362,6 +3444,60 @@ jsobj_print_jsarray(uintptr_t addr, jsobj_print_t *jsop)
 }
 
 static int
+jsobj_print_jstyped_array(uintptr_t addr, jsobj_print_t *jsop)
```
Are TypedArrays used for anything other than Buffers?  Would it be reasonable to have this code path print something like: "<TypedArray of length N>", rather than trying too hard to decode it?  Basically, as it is now, I'm not sure what this really _should_ do, but for the reasons you mentioned before, I'm reluctant to pretend like it has a "length" property that it doesn't really have.

misterdjules replied:
Typed arrays are used elsewhere, for instance to store [DomainFlags](https://github.com/nodejs/node/blob/master/src/node.cc#L940) and [tick info](https://github.com/nodejs/node/blob/master/src/node.cc#L969), but not for storing anything that is created by users, except for Buffers.

I would not display the length property for typed arrays with `::jsprint`, but maybe it would be useful to display the length as the output of `::nodebuffer` in case the buffer is a typed array? It seems handy to know how many bytes in memory to examine when looking at a buffer instance.

I would keep `jsobj_print_jstyped_array` printing other properties (e.g for use cases like `var buf = new Buffer('foo'); buf.some_property = true;`), but maybe that can be done by using `jsobj_print_jsobject` to print instances of JSTypedArray instance of keeping `jsobj_print_jstyped_array`.

What do you think?

davepacheco replied:
That seems reasonable, but so does doing nothing special (as long as there's another way to get the array data out).

Arrays can have named properties, too, and we don't print those.  We could, but then we couldn't print an Array the way it would look inline (since there's no place for other named properties), which is more convenient most of the time.  We could probably do better here more broadly, but I don't think it's been a major pain point.

misterdjules replied:
@davepacheco Sounds good, I hadn't seen your suggestion of "\<TypedArray of length N\>" initially, because to display a string within "\<\>" in GitHub comments, you need to escape "\<" and "\>" apparently. It definitely helps me to understand what you meant :) I updated your original comment, which is why we can see it now.

To summarize, I will just make `::jsprint` display "\<TypedArray of length N\>" for typed arrays.

davepacheco commented:
I've made a few inline comments.  In general, the debugger should present the facts, however ugly.  For that reason I prefer reporting the constructor as "Uint8Array", or maybe "Uint8Array (with Buffer prototype)".  Thanks!

misterdjules commented:
@davepacheco Thanks for the review, I replied to your comments and I will update this PR shortly.

I'll do a bit more research on how to get the actual prototype set on Buffer instances, and I'll keep you posted.

misterdjules commented:
@davepacheco I believe I have addressed all your comments.

I still need to do more research around how `Object.setPrototypeOf` affects V8's internal data structures, but I believe the current state of this PR already significantly improves support for node v4.x [1], and would be worth considering for a new release (pending the second review, obviously).

[1] Or rather for the next release of node v4.x, as some metadata is missing in node v4.0 and v4.1. See https://github.com/joyent/mdb_v8/issues/34 for more details.

misterdjules commented:
@davepacheco Regarding [1] above, I was mostly confused by the fact that [properties of functions' prototypes get a `constructor` property by default which is the function itself](https://github.com/nodejs/node/blob/master/deps/v8/src/factory.cc#L1301-L1327). It all makes sense now.

Now, regarding:

> In general, the debugger should present the facts, however ugly. For that reason I prefer reporting the constructor as "Uint8Array", or maybe "Uint8Array (with Buffer prototype)"

Even though I generally agree with the fact that the debugger should represent the facts, I still think it could be confusing for users who are used to use `::findjsobjects -c Buffer`, and who don't necessarily know that Buffer instances are now typed arrays, to have to use `::findjsobjects -c Uint8Array`.

The confusion might be made worse because `var buf = new Buffer(); console.log(buf.constructor.name);` outputs `'Buffer'`.

Thinking out loud: one potential solution would be to add another parameter to `::findjsobjects`, maybe "-C constructor_name" (capital c) to search for objects that have a `constructor` property whose name is `constructor_name`. Thus, we could use `::findjsobjects -C Buffer` with all versions of node to find Buffer instances.

At least, we should probably document/communicate around this change.

Do you mind if I open a separate issue to discuss potential solutions for this specific use case?

misterdjules commented:
Squashed commits into a single one and updated commit message + original PR description to match the commit message.

misterdjules commented:
@davepacheco I've actually ran the full tests suite again against all v0.10.40, v0.12.7, v4.0.0 and v4.1.0 node versions, 32 and 64 bits and I add [another commit](https://github.com/misterdjules/mdb_v8/commit/966bf63e2791463bc961f102ff2bea4c20b70134) that fixes what I believe is the last remaining issues.

If your second review is OK I'll squash these two commits into one.

davepacheco commented:
LGTM!

misterdjules commented:
@davepacheco Thank you! Squashed all commits into one, and made two minor changes:

1. Updated comment to not mention node and only V8 here: https://github.com/joyent/mdb_v8/pull/33/files#diff-38a6513060f0dadc2b8c2ea5c48a57d9R1052. Also updated V8 version number to reflect where the proper metadata landed (4.7): https://github.com/joyent/mdb_v8/pull/33/files#diff-38a6513060f0dadc2b8c2ea5c48a57d9R1063.

2. Use actual numbers to compare node version numbers: https://github.com/joyent/mdb_v8/pull/33/files#diff-1f7f4b61869c469eb7471735f5412183R123 and https://github.com/joyent/mdb_v8/pull/33/files#diff-1f7f4b61869c469eb7471735f5412183R197.

davepacheco commented:
Merged.  Thanks!  I'll wait for #37 and then issue a new release.
